### ISSUE TITLE
Sparse register kernels failing on Linux

### ISSUE THREAD

**Issue by FreddieWitherden** [libxsmm/libxsmm#599]
On an AARCH64 system (Neoverse 2):
```
~/libxsmm/samples/pyfr$ OMP_NUM_THREADS=1 ./pyfr_driver_asp_reg mats/p2/hex/m132-sp.mtx 48 100
CSR matrix data structure we just read:
rows: 27, columns: 81, elements: 216
LIBXSMM WARNING: timer is maybe not cycle-accurate!
computing golden solution...
...done!
computing libxsmm (A sparse) solution...
...done!
computing BLAS (A dense) solution...
...done!
max error beta=0 (libxmm vs. gold): 0.000000
max error beta=1 (libxmm vs. gold): 0.000000
max error beta=0 (dense vs. gold): 0.000000
max error beta=1 (dense vs. gold): 0.000000
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.000012
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 1.738403 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 17.601329 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=0): 3.476806
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.000012
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 1.735394 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 17.570867 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=1): 4.338486
time[s] MKL     (RM, M=27, N=48, K=81, beta=0): 0.000098
GFLOPS  MKL     (RM, M=27, N=48, K=81, beta=0): 2.152420
GB/s    MKL     (RM, M=27, N=48, K=81, beta=0): 0.531462
time[s] MKL     (RM, M=27, N=48, K=81, beta=1): 0.000097
GFLOPS  MKL     (RM, M=27, N=48, K=81, beta=1): 2.162731
GB/s    MKL     (RM, M=27, N=48, K=81, beta=1): 0.534008

LIBXSMM_VERSION: master-1.17-2204 (25692316)
AARCH64/DP    TRY    JIT    STA    COL
   0..13      0      0      0      0
  14..23      0      0      0      0
  24..64      2      2      0      0
Registry and code: 13 MB + 16 KB (gemm=2)
Uptime: 0.024736 s
```
which is a level of performance one would expect from a dense kernel (given the peak of the CPU is ~20 GFLOP/s double precision).  Analysis shows that the sparse kernels are failing to JIT and hence we are falling back to a dense kernel.  However, on an M1 Mac:
```
./pyfr_driver_asp_reg mats/p2/hex/m132-sp.mtx 48 100
CSR matrix data structure we just read:
rows: 27, columns: 81, elements: 216
computing golden solution...
...done!
computing libxsmm (A sparse) solution...
...done!
computing BLAS (A dense) solution...
...done!
max error beta=0 (libxmm vs. gold): 0.000000
max error beta=1 (libxmm vs. gold): 0.000000
max error beta=0 (dense vs. gold): 0.000000
max error beta=1 (dense vs. gold): 0.000000
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.000001
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 24.395294 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 247.002353 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=0): 48.790588
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.000001
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 23.298876 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 235.901124 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=1): 58.247191
time[s] MKL     (RM, M=27, N=48, K=81, beta=0): 0.000079
GFLOPS  MKL     (RM, M=27, N=48, K=81, beta=0): 2.668429
GB/s    MKL     (RM, M=27, N=48, K=81, beta=0): 0.658871
time[s] MKL     (RM, M=27, N=48, K=81, beta=1): 0.000074
GFLOPS  MKL     (RM, M=27, N=48, K=81, beta=1): 2.828779
GB/s    MKL     (RM, M=27, N=48, K=81, beta=1): 0.698464
```
which is in line with expectations.  Thus, the issue appears to be Linux specific.  I have confirmed that the CPU is being identified correctly, but am unsure where the code generation is failing.

**Comment by FreddieWitherden** [comment]
So I've done another pass and am still not entirely sure what is going on.

**Comment by alheinecke** [comment]
Hmm. I'm using an NV Jetson board and inserted some prints here:
https://github.com/libxsmm/libxsmm/blob/master/src/libxsmm_fsspmdm.c#L701

and I get:
```
~/Projects/LIBXSMM_workspace/libxsmm_github/samples/pyfr$ LIBXSMM_VERBOSE=-1 OMP_NUM_THREADS=6 ./pyfr_driver_asp_reg mats/p2/hex/m132-sp.mtx 48 1
CSR matrix data structure we just read:
rows: 27, columns: 81, elements: 216
LIBXSMM WARNING: timer is maybe not cycle-accurate!
LIBXSMM-JIT-DUMP(ptr:file) 0x7fa8d6e000 : libxsmm_aarch64_f64_nn_27x2x81_0_48_48_a1_b0_p0.sreg
LIBXSMM-JIT-DUMP(ptr:file) 0x7fa8d6f000 : libxsmm_abi1_aarch64_f64_nn_8x27x81_48_81_48_a1_b0_p0_br0_uh0_si0_tc-abid_avnni0_bvnni0_cvnni0_decompress_A0_spfactor1.mxm
LIBXSMM-JIT-DUMP(ptr:file) 0x7fa8d6c000 : libxsmm_aarch64_f64_nn_27x2x81_0_48_48_a1_b1_p0.sreg
LIBXSMM-JIT-DUMP(ptr:file) 0x7fa8d6d000 : libxsmm_abi1_aarch64_f64_nn_8x27x81_48_81_48_a1_b1_p0_br0_uh0_si0_tc-abid_avnni0_bvnni0_cvnni0_decompress_A0_spfactor1.mxm
computing golden solution...
...done!
computing libxsmm (A sparse) solution...
sparse
sparse
...done!
computing BLAS (A dense) solution...
...done!
max error beta=0 (libxmm vs. gold): 0.000000
max error beta=1 (libxmm vs. gold): 0.000000
max error beta=0 (dense vs. gold): 0.000000
max error beta=1 (dense vs. gold): 0.000000
sparse
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.000548
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.037833 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.383055 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.075665
sparse
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.000920
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.022534 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.228152 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.056334
time[s] MKL     (RM, M=27, N=48, K=81, beta=0): 0.000237
GFLOPS  MKL     (RM, M=27, N=48, K=81, beta=0): 0.886378
GB/s    MKL     (RM, M=27, N=48, K=81, beta=0): 0.218859
time[s] MKL     (RM, M=27, N=48, K=81, beta=1): 0.000279
GFLOPS  MKL     (RM, M=27, N=48, K=81, beta=1): 0.752406
GB/s    MKL     (RM, M=27, N=48, K=81, beta=1): 0.185779

LIBXSMM_VERSION: master-1.17-2219 (25692331)
AARCH64/DP    TRY    JIT    STA    COL
   0..13      0      0      0      0
  14..23      0      0      0      0
  24..64      2      2      0      0
Registry and code: 13 MB + 16 KB (gemm=2)
Uptime: 0.011032 s
```

**Comment by alheinecke** [comment]
When I run on a i7-11700K (AVX512, one unit, 4.8 GHz, ~153.6 GFLOPS):
```
 time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.000003
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 6.706181 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 67.900087 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=0): 13.412363
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.000003
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 7.339595 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 74.313399 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=1): 18.348987
```
Multi-core on the Rocket Lake (N=384):
```
time[s] LIBXSMM (RM, M=27, N=384, K=81, beta=0): 0.000004
GFLOPS  LIBXSMM (RM, M=27, N=384, K=81, beta=0): 39.744625 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=384, K=81, beta=0): 402.414330 (dense)
GB/s    LIBXSMM (RM, M=27, N=384, K=81, beta=0): 79.489250
time[s] LIBXSMM (RM, M=27, N=384, K=81, beta=1): 0.000004
GFLOPS  LIBXSMM (RM, M=27, N=384, K=81, beta=1): 40.402133 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=384, K=81, beta=1): 409.071593 (dense)
GB/s    LIBXSMM (RM, M=27, N=384, K=81, beta=1): 101.005332
```

And on the Jetson Xavier NX board:
```
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.000022
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 0.959371 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=0): 9.713634 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=0): 1.918742
time[s] LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.000022
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 0.961060 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=48, K=81, beta=1): 9.730736 (dense)
GB/s    LIBXSMM (RM, M=27, N=48, K=81, beta=1): 2.402651
```
and 6 cores (N=288) on the Xavier:
```
time[s] LIBXSMM (RM, M=27, N=288, K=81, beta=0): 0.000029
GFLOPS  LIBXSMM (RM, M=27, N=288, K=81, beta=0): 4.318919 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=288, K=81, beta=0): 43.729059 (dense)
GB/s    LIBXSMM (RM, M=27, N=288, K=81, beta=0): 8.637839
time[s] LIBXSMM (RM, M=27, N=288, K=81, beta=1): 0.000029
GFLOPS  LIBXSMM (RM, M=27, N=288, K=81, beta=1): 4.342700 (sparse)
GFLOPS  LIBXSMM (RM, M=27, N=288, K=81, beta=1): 43.969837 (dense)
GB/s    LIBXSMM (RM, M=27, N=288, K=81, beta=1): 10.856750
```
and this Stream on the Xavier... not sure why it's not getting close to BW with sparse reg kernels:
```
Function    Best Rate MB/s  Avg time     Min time     Max time
Copy:           37682.6     0.031560     0.031305     0.032296
Scale:          38208.4     0.031099     0.030874     0.031618
Add:            29435.7     0.060526     0.060113     0.069795
Triad:          28818.2     0.062205     0.061401     0.062629
```

**Comment by FreddieWitherden** [comment]
Running on a Neoverse 2:
```
~/Projects/LIBXSMM_workspace/libxsmm_github/samples/pyfr$ LIBXSMM_VERBOSE=-1 OMP_NUM_THREADS=6 ./pyfr_driver_asp_reg mats/p2/hex/m132-sp.mtx 48 1
CSR matrix data structure we just read:
rows: 27, columns: 81, elements: 216
LIBXSMM WARNING: timer is maybe not cycle-accurate!
LIBXSMM-JIT-DUMP(ptr:file) 0x7fa8d6e000 : libxsmm_aarch64_f64_nn_27x2x81_0_48_48_a1_b0_p0.sreg
LIBXSMM-JIT-DUMP(ptr:file) 0x7fa8d6f000 : libxsmm_abi1_aarch64_f64_nn_8x27x81_48_81_48_a1_b0_p0_br0_uh0_si0_tc-abid_avnni0_bvnni0_cvnni0_decompress_A0_spfactor1.mxm
LIBXSMM-JIT-DUMP(ptr:file) 0x7fa8d6c000 : libxsmm_aarch64_f64_nn_27x2x81_0_48_48_a1_b1_p0.sreg
LIBXSMM-JIT-DUMP(ptr:file) 0x7fa8d6d000 : libxsmm_abi1_aarch64_f64_nn_8x27x81_48_81_48_a1_b1_p0_br0_uh0_si0_tc-abid_avnni0_bvnni0_cvnni0_decompress_A0_spfactor1.mxm
[...]
```
which is the same output as you.  However, this is problematic in the sense the only sparse register kernel we're seeing is two wide (one vector length).  The only time we don't try and JIT a two wide kernel is if the first fails to JIT as per:

https://github.com/libxsmm/libxsmm/blob/master/src/libxsmm_fsspmdm.c#L178

Sticking a print just above here and re-running:
```
LIBXSMM_VERBOSE=-1 OMP_NUM_THREADS=6 ./pyfr_driver_asp_reg mats/p2/hex/m132-sp.mtx48 1
CSR matrix data structure we just read:
rows: 27, columns: 81, elements: 216
LIBXSMM WARNING: timer is maybe not cycle-accurate!
LIBXSMM-JIT-DUMP(ptr:file) 0xffff8b1cd000 : libxsmm_aarch64_f64_nn_27x2x81_0_48_48_a1_b0_p0.sreg
k_sparse1: (nil)
LIBXSMM-JIT-DUMP(ptr:file) 0xffff8b1ce000 : libxsmm_abi1_aarch64_f64_nn_8x27x81_48_81_48_a1_b0_p0_br0_uh0_si0_tc-abid_avnni0_bvnni0_cvnni0_decompress_A0_spfactor1.mxm
LIBXSMM-JIT-DUMP(ptr:file) 0xffff8b1cb000 : libxsmm_aarch64_f64_nn_27x2x81_0_48_48_a1_b1_p0.sreg
k_sparse1: (nil)
LIBXSMM-JIT-DUMP(ptr:file) 0xffff8b1cc000 : libxsmm_abi1_aarch64_f64_nn_8x27x81_48_81_48_a1_b1_p0_br0_uh0_si0_tc-abid_avnni0_bvnni0_cvnni0_decompress_A0_spfactor1.mxm
```
So while a file is getting written out to disk, we're getting a NULL pointer back in `ksparse_1`.

**Comment by FreddieWitherden** [comment]
I am also observing similar issues on an x86 AVX-512 machine.  With the same sparse pointer dumping patch:
```
LIBXSMM_VERBOSE=-1 ./pyfr_driver_asp_reg ./mats/p1/tri/m6-sp.mtx 480 10000
CSR matrix data structure we just read:
rows: 6, columns: 6, elements: 24
LIBXSMM-JIT-DUMP(ptr:file) 0x7f712680c000 : libxsmm_clx_f64_nn_6x8x6_0_480_480_a1_b0_p0.sreg
We got: (nil)
LIBXSMM-JIT-DUMP(ptr:file) 0x7f712680d000 : libxsmm_abi1_clx_f64_nn_8x6x6_480_6_480_a1_b0_p0_br0_uh0_si0_tc-abid_avnni0_bvnni0_cvnni0_decompress_A0_spfactor1.mxm
LIBXSMM-JIT-DUMP(ptr:file) 0x7f712680a000 : libxsmm_clx_f64_nn_6x8x6_0_480_480_a1_b1_p0.sreg
We got: (nil)
LIBXSMM-JIT-DUMP(ptr:file) 0x7f712680b000 : libxsmm_abi1_clx_f64_nn_8x6x6_480_6_480_a1_b1_p0_br0_uh0_si0_tc-abid_avnni0_bvnni0_cvnni0_decompress_A0_spfactor1.mxm
computing golden solution...
...done!
computing libxsmm (A sparse) solution...
...done!
computing BLAS (A dense) solution...
...done!
max error beta=0 (libxmm vs. gold): 0.000000
max error beta=1 (libxmm vs. gold): 0.000000
max error beta=0 (dense vs. gold): 0.000000
max error beta=1 (dense vs. gold): 0.000000
```
Where although the kernel is being JIT'ed and written out to disk, a null pointer is being returned from `libxsmm_create_spgemm_csr_areg_v2`.

**Comment by alheinecke** [comment]
OK, I put two markers:
after code gen:
https://github.com/libxsmm/libxsmm/blob/master/src/libxsmm_main.c#L2203
before exit
https://github.com/libxsmm/libxsmm/blob/master/src/libxsmm_main.c#L2284

and got:
```
 after code-gen 0x0
LIBXSMM-JIT-DUMP(ptr:file) 0x7fc86d1c7000 : libxsmm_clx_f64_nn_27x8x81_0_48_48_a1_b0_p0.sreg
 before exit 0xFFFFFFFFFFFFFFFF
```

@hfp: can you please have a look, may be this is due to the data insertion?

**Comment by FreddieWitherden** [comment]
I think the issue might be here:

https://github.com/libxsmm/libxsmm/blob/master/src/libxsmm_main.c#L2244

as I do not think that the resulting pointer will be page aligned which may cause protection issues.  See 

https://github.com/libxsmm/libxsmm/blob/master/src/generator_x86_instructions.c#L4529

as it is the data size which includes the padding required such that the data starts on a page sized boundary.

**Comment by alheinecke** [comment]
yeah, and also the padding spaces doesn't seem to be set to zero. Perhaps some merges have been messed up?!

I suggest the following:
0-code_size -> X
code_size - next page boundary -> set to 0, X
next page boundary - data end -> R

we might need to track the padding space in io_generated code.

**Comment by alheinecke** [comment]
@FreddieWitherden : why is an alignment needed here: https://github.com/libxsmm/libxsmm/blob/master/src/generator_x86_instructions.c#L4483 ? 

We know by construction that the data later will start page  aligned.

I'm starting to feel that the issue was always there and was introduced when we move the buffer from the beginning to the end of the code?!

**Comment by FreddieWitherden** [comment]
> @FreddieWitherden : why is an alignment needed here: https://github.com/libxsmm/libxsmm/blob/master/src/generator_x86_instructions.c#L4483 ?
> 
> We know by construction that the data later will start page aligned.
> 
> I'm starting to feel that the issue was always there and was introduced when we move the buffer from the beginning to the end of the code?!

Consider adding one byte of data (a character) to the data segment, and then adding a 512-bit vector.  If we did not do anything here, the vector would be at (page_aligned_addr + 1) and thus would be badly out of alignment.

I think the data code is correct, it is just we're not correctly computing the address of where the data starts (it is base + (code_size aligned up to the nearest page) rather than (base + code_size).  I believe the padding data is currently garbage.

**Comment by alheinecke** [comment]
I see, now I get this add data part. However, wouldn't it be better to say that each new section that is submitted to the is CL-aligned. Then calling functions don't need to compute an alignment?!

**Comment by FreddieWitherden** [comment]
> I see, now I get this add data part. However, wouldn't it be better to say that each new section that is submitted to the is CL-aligned. Then calling functions don't need to compute an alignment?!

In the A-not-in-registers case on AARCH64 we build up the data segment in the main sequencing loop:

https://github.com/libxsmm/libxsmm/blob/master/src/generator_spgemm_csr_asparse_reg.c#L1108

with each call adding a single float/double to the segment.  If all segments were CL-aligned we'd need to append the data in one big chunk which would require an additional pass over the operations (as what actually needs to added to the segment is a subset of A, rather than the entire matrix or just its unique values).  The current API makes it easy to figure out what we need to add as we're going along.

**Comment by alheinecke** [comment]
I tested your fix (#617) an the problem is gone. Can we close this issue as well?

**Comment by FreddieWitherden** [comment]
Yes, assuming everyone is happy with the structure of the fix everything looks good to me.
